### DISCUSSION TITLE
Background plugins

### DISCUSSION THREAD

**Discussion by idanarye** [nushell/nushell#12849] [Ideas]
Plugins currently can only read commands from the shell and respond to these command immediately. I think they could be so much more powerful if they could also _initiate_ interaction with the shell when something happen on their side - when they receive a message on a socket they've opened, or when something they watch changes, or even just when a timer expires.

The language that the plugin is written in can be the one responsible for running the background task (in Rust's case, though, [the official `nu-plugin` crate](https://crates.io/crates/nu-plugin) should assist with that). What the nushell protocol needs to support is for the plugin itself to initiate things. Namely:

* Change variables. For example, a plugin could change some field in `$env` which the prompt uses.
* Speaking of the prompt - force a prompt update.
* Run a shell command. Or, to make this more sanitized, separate this into two
  * Parse and compile a closure (signature of the closure provided programmatically, so the string passed will just be the closure's body)
  * Run a closure with arguments.
* Set the actual command line without sending it (so that they user can edit it). Though this may be a bit dangerous.

**Comment by ghost** [comment]
> they receive a message on a socket they've opened

Im not sure what exactly you want to achieve with this to be honest. a specific use case would be helpful for me.

a few request-sources and better alternatives i can think of:

* shells are temporary things, therefore things like a IRCd dont make sense.
*  for temporary webservers, etc i use [nq](https://github.com/leahneukirchen/nq) or just GNU-screen since its persists post-crash, etc
* theres no reason for a background ssh session to notify you about anything
* if your desktop does not support libnotify you could just use [tiramisu](https://github.com/Sweets/tiramisu) (also scriptable) or dunst

a few results it could achieve and why i think theyre a bad idea:

* anything live change is probably bad
  * multithreading in rust is delicate and this would be hard to do without using `unsafe` i think. and well running a command while the user is typing text IS multithreading.
  * changing how things work while the user is doing something is confusing. shortcuts are great, but  dont want to get random new aliases while typing because someone pinged my webserver.
  * prompt changes while typing are either called tab-completion or would only confuse the hell out of me.
* just adding things to stdout is pretty annoying (just try working in the linux-tty while a driver has issues). and this is also already possible by just starting something else paralel to nushell (just replace your terminal launch with something like `xterm -e bash -c "recieve_notifications_program & nu"`)
* getting new data into the prompt is always a data-pull and things like FIFO are already well established for this purpose and also make it easier to share the data between all shells

another idea would be a application, which uses nushell as its interface.  
you can look at couchbase-shell. i think someone created a repo for making this easier at some point, but i can't find it right now.


PS: i wrote another gigantic wall of text.. i don't want / mean to directly shoot you down with this. please don't feel discouraged due to this - i just tend to think way to much about such things and often don't see the obvious things.

**Reply by idanarye** [reply]
The actual use case I had in mind is for a live terminal session that gets commands from my editor via a UNIX socket. I currently use GNU screen for that, but it has some limitations:

1. I can launch the command, but I can't tell programmatically when exactly it finishes.
2. I can't read the command output from the editor (I can't read the screen log, but it's hard to tell when the output starts and where it ends. Plus it has ANSI symbols)
3. Writing output to the command is tricky - I can't tell programmatically if it'll go into the shell or into the process, because I don't know when the process has finished.
4. In a similar fashion, I can't tell what the exit status is.
5. If I want to use the same shell myself (part of the point is being able to use it) I keep tripping over the screen shortcut (CTRL+a)

I thought, if I can have a nushell plugin that can listen to a socket and run commands, that would solve all these problems...

I think at least the first four points can be done today if I modify the command, but that would be a problem with history since the modified command will end up there. Does normal nushell have any way to run one command, but have another get written to history?

**Reply by ghost** [reply]
so you want a REPL similar to lua / python / bash / ..?  
i want that too, but well well see when it gets added

**Comment by likern** [comment]
@Jan9103 My use case is to want to create http service (server) which will listen on some endpoint and being able to be triggered by external webhook.

In my exact use case it's s bot, webhook is initiated (called) by gitlab.

**Reply by ghost** [reply]
i also tested a bit of things with using nu as a webserver. probably the best approach right now would be to use netcat within nushell to have nushell as the controling program and loop over requests.  
something like this might be adaptable <https://dev.to/leandronsp/building-a-web-server-in-bash-part-i-sockets-2n8b>

**Reply by ghost** [reply]
Hi,  
i know its been a while, but ive sat down and created a library for this usecase since i wanted something similar for a long time: <https://github.com/Jan9103/webserver.nu>

**Reply by likern** [reply]
Amazing. I will try it out.
